src/xml_utils.cpp

```diff
@@ -0,0 +1,1850 @@
+      pugi::xpath_node node = *it;
+      for (pugi::xml_node temp: node.node().children()) {
+        std::string name = temp.attribute("name").as_string();
+        if (name != "") {
+          names.push_back(name);
+          ids.push_back(node.node().attribute("id").as_string());
+          version.push_back(node.node().attribute("version").as_string());
+        }
+      }
+    }
+  }
+
+  output.push_back(names);
+  output.push_back(ids);
+  output.push_back(version);
+
+  return output;
+}
+
+
+
+
+std::list<std::vector<std::string>> xml_utils::mzxml_software_parser(const pugi::xml_node& node_mzxml) {
+
+  std::list<std::vector<std::string>> output;
+  std::vector<std::string> names;
+  std::vector<std::string> ids;
+  std::vector<std::string> version;
+
+  std::string search_software = "//msInstrument/child::node()[starts-with(name(), 'soft')]";
+  pugi::xpath_node_set xps_software = node_mzxml.select_nodes(search_software.c_str());
+
+  if (xps_software.size() > 0) {
+    for (pugi::xpath_node_set::const_iterator it = xps_software.begin(); it != xps_software.end(); ++it) {
+      pugi::xpath_node node = *it;
+      names.push_back(node.node().attribute("name").as_string());
+      ids.push_back(node.node().attribute("type").as_string());
+      version.push_back(node.node().attribute("version").as_string());
+    }
+  }
+
+  output.push_back(names);
+  output.push_back(ids);
+  output.push_back(version);
+
+  return output;
+}
+
+
+
+
+Rcpp::List xml_utils::parse_software(const pugi::xml_node& root) {
+
+  Rcpp::List software_list;
+
+  std::list<std::vector<std::string>> output;
+
+  const char* root_name = root.name();
+
+  if (strcmp("indexedmzML", root_name) == 0) {
+    output = xml_utils::mzml_software_parser(root.child("mzML"));
+
+  } else if (strcmp("mzXML", root_name) == 0) {
+    output = xml_utils::mzxml_software_parser(root);
+
+  } else {
+    std::cout << "\u2717 The file must be in valid mzML or mzXML format!" << std::endl;
+    return software_list;
+  }
+
+  auto it2 = output.begin();
+  software_list["name"] = Rcpp::wrap(*it2);
+  std::advance(it2, 1);
+  software_list["id"] = Rcpp::wrap(*it2);
+  std::advance(it2, 1);
+  software_list["version"] = Rcpp::wrap(*it2);
+  software_list.attr("class") = Rcpp::CharacterVector::create("data.table", "data.frame");
+
+  return software_list;
+}
+
+
+
+
+void xml_utils::mzml_spectra_headers_parser(const pugi::xml_node& node_mzml, xml_utils::spectraHeaders& output) {
+
+  std::string search_run = "//run";
+
+  pugi::xpath_node xps_run = node_mzml.select_node(search_run.c_str());
+
+  pugi::xml_node spec_list = xps_run.node().child("spectrumList");
+
+  if (spec_list != NULL) {
+
+    std::vector<pugi::xml_node> spectra;
+
+    for (pugi::xml_node child = spec_list.first_child(); child; child = child.next_sibling()) {
+      spectra.push_back(child);
+    }
+
+    int number_spectra = spectra.size();
+
+    output.index.resize(number_spectra);
```

Comment:
I'd say to either make an utility function that does all the resizing or convert `spectraHeader` to a `class` and let a member function do it.

Yet another option is to specify the size when constructing `spectraHeader`. In that case  this function should construct the variable and return it.